kernel/sched.c

```diff
@@ -1117,7 +1117,9 @@ static int32_t z_tick_sleep(k_ticks_t ticks)
 
 	__ASSERT(!z_is_thread_state_set(arch_current_thread(), _THREAD_SUSPENDED), "");
 
-	ticks = (k_ticks_t)expected_wakeup_ticks - sys_clock_tick_get_32();
+	uint32_t left_ticks = expected_wakeup_ticks - sys_clock_tick_get_32();
+
+	ticks = (k_ticks_t)(int32_t)left_ticks;
```

Comment:
Sorry, "requires a 32 bit **unsigned** subtraction", that is.

Reply:
Thank you @andyross, I have to add more explanation.

I think there should be `(int32_t)` casting, because `sys_clock_tick_get_32()` can exceed `expected_wakeup_ticks` in the timeout case.

Case 1. `sys_clock_tick_get_32()` equals `expected_wakeup_ticks` after wakeup.
It may be timeout case without `k_wakeup()` call.

```
ticks = expected_wakeup_ticks - sys_clock_tick_get_32()  /* maybe 0, no problem */
```

Case 2. `sys_clock_tick_get_32()` exceeds `expected_wakeup_ticks` after wakeup.
The `k_sleep()` caller thread can be suspended by higher priority threads  after wakeup.

```
ticks = expected_wakeup_ticks - sys_clock_tick_get_32()  /* maybe very very big positive number */
```

Solution:

```C
uint32_t left_ticks = expected_wakeup_ticks - sys_clock_tick_get_32();  /* first, make a 32 bit unsigned subtraction */
int32_t signed_left_ticks = (int32_t)left_ticks;  /* for handling the negative value correctly */
ticks = (k_ticks_t)signed_left_ticks;  /* force return value type */
```

I put the last casting for avoiding implicit type conversion, and this PR version is intended to highlight `a 32 bit unsigned subtraction`. If I write down the code in the simplest way:

```C
ticks = (int32_t)(expected_wakeup_ticks - sys_clock_tick_get_32());
```

Reply:
@andyross May I put some comments to the code?

```C
        /* We require a 32 bit unsigned subtraction to care a wraparound */
	uint32_t left_ticks = expected_wakeup_ticks - sys_clock_tick_get_32();

        /* To handle a negative value correctly, once type-cast it to signed 32 bit */
	ticks = (k_ticks_t)(int32_t)left_ticks;
```